### PR TITLE
Typescript Conversion

### PR DESCRIPTION
**Pull request by patburke234**
Convert to Typescript
Add new UI controls

### CODE REVIEW

**Comment by AdrianMachado** [comment]
Will review this weekend.

**Comment by AdrianMachado** [comment]
Is it possible to split this out into multiple Pull requests to make this easier to review? At least splitting out the new features from converting existing features would be helpful

File: src/components/ArrowListItem/ArrowListItem.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      softKeyManager
+    } = props;
+
+    const handleFocusChange = isNowFocused => {
+      if (isNowFocused) {
+        if(onFocusChange)
+          onFocusChange(index);
+        if(onClick)
+          softKeyManager.setCenterCallback(onClick);
+      }
```
Comment (AdrianMachado) [inline src/components/ArrowListItem/ArrowListItem.tsx:41]: Use curly braces rather than just tabbing

File: src/components/Audio/Audio.scss [hunk 2]
```diff
@@ -0,0 +1,4 @@
+.audio-softkey-container{
+    margin-top: 5px;
+    margin-left: 5px;
+}
```
Comment (AdrianMachado) [inline src/components/Audio/Audio.scss:4]: Should have newline

File: src/components/Audio/Audio.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+interface LocalState {
+    play: boolean;
+}
+
+class AudioComponent extends React.Component<Props, LocalState> {
+    private audioElem: HTMLAudioElement;
+    constructor(props) {
+        super(props);
+        this.state = {
+            play: false
```
Comment (AdrianMachado) [inline src/components/Audio/Audio.tsx:21]: shouldn't this be `playing` or `isPlaying` rather than just `play`?

File: src/components/Audio/Audio.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        };
+        this.audioElem = new Audio(this.props.url);
+        this.audioElem.onended = this.stop;
```
Comment (AdrianMachado) [inline src/components/Audio/Audio.tsx:24]: typo or is this function actually not camel case?

File: src/components/Audio/Audio.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        this.setState({ play: true });
+        this.audioElem.play();
+    }
+    
+    pause = () => {
+        this.setState({ play: false });
+        this.audioElem.pause();
+    }
+    
+    stop = () => {
```
Comment (AdrianMachado) [inline src/components/Audio/Audio.tsx:37]: would prefer if these function were suffixed with `Playback` so `startPlayback`, `pausePlayback`, `stopPlayback`. Easier to understand imo

File: src/components/BodyTextListItem/BodyTextListItem.tsx [hunk 6]
```diff
@@ -1,13 +1,21 @@
...
 import './BodyTextListItem.scss';
 
 const prefixCls = 'kai-btl';
 
-const BodyTextListItem = React.memo(
+interface LocalProps {
+  header: string;
+  body?: string;
+  focusColor?:string;
+  forwardedRef?:any;
```
Comment (AdrianMachado) [inline src/components/BodyTextListItem/BodyTextListItem.tsx:13]: Is there no equivalent to the propType I used

File: src/components/Camera/Camera.scss [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    text-align: center;
+        
+    .kai-icon-button-input{
+        background: #d0d0d0;
+        padding: 20px;
+        border-radius: 50px;
+    }
+}
+
+.hidden{ display: none;}
```
Comment (AdrianMachado) [inline src/components/Camera/Camera.scss:22]: newline

File: src/components/Camera/Camera.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
+import React from 'react';
```
Comment (AdrianMachado) [inline src/components/Camera/Camera.tsx:1]: can we split this camera stuff into another PR? Same with the Audio stuff. Appreciate you making these though!

File: src/components/DateSelector/DateSelector.tsx [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+  onOK: (res?: Date|null) => void;
+  onCancel: () => void;
+  header: React.ElementType | string | undefined;
+  inputOptions: any;
+  minDate: Date;
+  maxDate: Date;
+  initialDate: Date;
+}
+
+const Months = {
```
Comment (AdrianMachado) [inline src/components/DateSelector/DateSelector.tsx:20]: Should we `Object.freeze` this since it's an enum?

File: src/components/DateSelector/DateSelector.tsx [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+  years: number[],
+  months: string[],
+  days: number[],
+  selectedMonth: number;
+  selectedDay: number;
+  selectedYear: number;
+}
+
+class DateSelector extends React.Component<Props, LocalState> {
+    private input: HTMLInputElement | null = null
```
Comment (AdrianMachado) [inline src/components/DateSelector/DateSelector.tsx:51]: Indentation seems off

File: src/components/DateSelector/DateSelector.tsx [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+      selectedYear: this.props.initialDate.getFullYear(),
+      isLoading: false
+    });
+  }
+
+  focusLast() {
+  }
+
+  focus() {
+  }
```
Comment (AdrianMachado) [inline src/components/DateSelector/DateSelector.tsx:100]: Do we need these?

File: src/components/DateSelector/DateSelector.tsx [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    ReactDOM.unmountComponentAtNode(div);
+    document.body.removeChild(div);
+  }
+
+  render({ ...config, close: closeWindow });
+}
+
+
+export {showDateSelctor};
+export default DateSelector;
```
Comment (AdrianMachado) [inline src/components/DateSelector/DateSelector.tsx:224]: newline

File: src/components/Dialog/Dialog.tsx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    this.lastFocus = null;*/
+  }
+
+  focus() {
+    /*if (this.input) {
+      this.input.focus();
+    } else if(this.el) {
+      this.el.focus();
+    }*/
+  }
```
Comment (AdrianMachado) [inline src/components/Dialog/Dialog.tsx:46]: Do we need these if its commented out?

File: src/components/Dialog/Dialog.tsx [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+        break;
+
+      default:
+        break;
+    }
+  };
+
+  close() {
+    this.props.close();
+    //this.focusLast();
```
Comment (AdrianMachado) [inline src/components/Dialog/Dialog.tsx:77]: leftover comment

File: src/components/Dialog/Dialog.tsx [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  render() {
+    const { header, content, type, inputOptions, onOK, onClose, close } = this.props;
+    return (
+      <div className="systemContent">
+        <div
+          //ref={node => { this.el = node }}
```
Comment (AdrianMachado) [inline src/components/Dialog/Dialog.tsx:85]: leftover comment

File: src/components/Audio/Audio.tsx (hunk 4 above)
Comment (patburke234) [reply src/components/Audio/Audio.tsx:24]: This is the way it is defined.

File: src/components/Audio/Audio.tsx (hunk 5 above)
Comment (patburke234) [reply src/components/Audio/Audio.tsx:37]: Sure, makes sense

File: src/components/Audio/Audio.tsx (hunk 3 above)
Comment (patburke234) [reply src/components/Audio/Audio.tsx:21]: Done

File: src/components/Audio/Audio.scss (hunk 2 above)
Comment (patburke234) [reply src/components/Audio/Audio.scss:4]: Done

File: src/components/ArrowListItem/ArrowListItem.tsx (hunk 1 above)
Comment (patburke234) [reply src/components/ArrowListItem/ArrowListItem.tsx:41]: Done, added tslint to enforce this

File: src/components/DateSelector/DateSelector.tsx (hunk 11 above)
Comment (patburke234) [reply src/components/DateSelector/DateSelector.tsx:100]: Nope, removed.

File: src/components/DateSelector/DateSelector.tsx (hunk 12 above)
Comment (patburke234) [reply src/components/DateSelector/DateSelector.tsx:224]: Done

File: src/components/Dialog/Dialog.tsx (hunk 13 above)
Comment (patburke234) [reply src/components/Dialog/Dialog.tsx:46]: Removed

File: src/components/Dialog/Dialog.tsx (hunk 15 above)
Comment (patburke234) [reply src/components/Dialog/Dialog.tsx:85]: Removed

File: src/views/TriColumnListView/TriColumnListView.tsx [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+      }
+    };
+
+    
+    const handleKeyDown = useCallback(
+      e => {
+        let index = activeTab;
+        switch (e.key) {
+          case 'ArrowLeft':
+            index = index - 1 >= 0 ? index - 1 : 3 - 1;
```
Comment (AdrianMachado) [inline src/views/TriColumnListView/TriColumnListView.tsx:57]: I am guessing you wrote 3 - 1 instead of 2 on purpose?

File: src/utils/dates.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import { Time } from "../components/TimeSelector/TimeSelector";
+
+export const getDaysInMonth = (month,year) => {
+    return new Date(year, month, 0).getDate();
+  };
+
+export const get12Hour = (hh) => {
```
Comment (AdrianMachado) [inline src/utils/dates.ts:7]: I usually prefer not putting numbers in variable names, can you make this `getTwelveHour`?

File: src/react-app-env.d.ts [hunk 18]
```diff
@@ -0,0 +1 @@
+/// <reference types="react-scripts" />
```
Comment (AdrianMachado) [inline src/react-app-env.d.ts:1]: What's this for?

**Comment by AdrianMachado** [comment]
Pls fix the merge conflicts and add new lines at the end of all files

File: src/components/BodyTextListItem/BodyTextListItem.tsx (hunk 6 above)
Comment (patburke234) [reply src/components/BodyTextListItem/BodyTextListItem.tsx:13]: Not that I could find, no. If you find one, happy to replace it.

File: src/components/DateSelector/DateSelector.tsx (hunk 9 above)
Comment (patburke234) [reply src/components/DateSelector/DateSelector.tsx:20]: Good call, missed this.

File: src/components/DateSelector/DateSelector.tsx (hunk 10 above)
Comment (patburke234) [reply src/components/DateSelector/DateSelector.tsx:51]: Fixed

File: src/views/TriColumnListView/TriColumnListView.tsx (hunk 16 above)
Comment (patburke234) [reply src/views/TriColumnListView/TriColumnListView.tsx:57]: Yeah... totally. Haha. I think it was just to illustrate the -1 but doesn't make sense in hindsight.

File: src/utils/dates.ts (hunk 17 above)
Comment (patburke234) [reply src/utils/dates.ts:7]: Sure, done

File: src/react-app-env.d.ts (hunk 18 above)
Comment (patburke234) [reply src/react-app-env.d.ts:1]: Weird, looks auto-generated. Not needed.

**Comment by patburke234** [comment]
> Pls fix the merge conflicts and add new lines at the end of all files

Why newlines at the end of all files?

**Comment by patburke234** [comment]
Doesn't let me merge above since I don't have write access. Can you grant that to me? You can remove it after if you wish.

**Comment by patburke234** [comment]
git checkout -b patburke234-master master
git pull https://github.com/patburke234/KaiUI-Typescript.git master
git checkout master
git merge --no-ff patburke234-master
git push origin master

**Comment by AdrianMachado** [comment]
See https://stackoverflow.com/questions/15376241/why-does-git-say-my-master-branch-is-already-up-to-date-even-though-it-is-not/15381342
I think you need to fetch from the master branch on this repo, not your own, then merge.

**Comment by patburke234** [comment]
Created a new one
